src/model/OutputData.ts

```diff
@@ -0,0 +1,163 @@
+import {
+	MintKeys,
+	Proof,
+	SerializedBlindedMessage,
+	SerializedBlindedSignature,
+	SerializedDLEQ
+} from './types';
+import {
+	blindMessage,
+	constructProofFromPromise,
+	serializeProof
+} from '@cashu/crypto/modules/client';
+import { BlindedMessage } from './BlindedMessage';
+import { bytesToHex, hexToBytes, randomBytes } from '@noble/hashes/utils';
+import { DLEQ, pointFromHex } from '@cashu/crypto/modules/common';
+import { verifyDLEQProof_reblind } from '@cashu/crypto/modules/client/NUT12';
+import { bytesToNumber, numberToHexPadded64, splitAmount } from '../utils';
+import { deriveBlindingFactor, deriveSecret } from '@cashu/crypto/modules/client/NUT09';
+
+export interface OutputDataLike {
+	blindedMessage: SerializedBlindedMessage;
+	blindingFactor: bigint;
+	secret: Uint8Array;
+
+	toProof: (signature: SerializedBlindedSignature, keyset: MintKeys) => Proof;
+}
+
+export type OutputDataFactory = (amount: number, keys: MintKeys) => OutputDataLike;
+
+export function isOutputDataFactory(
+	value: Array<OutputData> | OutputDataFactory
+): value is OutputDataFactory {
+	return typeof value === 'function';
+}
+
+export class OutputData implements OutputDataLike {
+	blindedMessage: SerializedBlindedMessage;
+	blindingFactor: bigint;
+	secret: Uint8Array;
+
+	constructor(blindedMessage: SerializedBlindedMessage, blidingFactor: bigint, secret: Uint8Array) {
+		this.secret = secret;
+		this.blindingFactor = blidingFactor;
+		this.blindedMessage = blindedMessage;
+	}
+
+	toProof(sig: SerializedBlindedSignature, keyset: MintKeys) {
+		const dleq =
+			sig.dleq == undefined
+				? undefined
+				: ({
+						s: hexToBytes(sig.dleq.s),
+						e: hexToBytes(sig.dleq.e),
+						r: this.blindingFactor
+				  } as DLEQ);
+		const blindSignature = {
+			id: sig.id,
+			amount: sig.amount,
+			C_: pointFromHex(sig.C_),
+			dleq: dleq
+		};
+		const A = pointFromHex(keyset.keys[this.blindedMessage.amount]);
+		const proof = constructProofFromPromise(blindSignature, this.blindingFactor, this.secret, A);
+		const serializedProof = {
+			...serializeProof(proof),
+			...(dleq && {
+				dleqValid: verifyDLEQProof_reblind(this.secret, dleq, proof.C, A)
+			}),
+			...(dleq && {
+				dleq: {
+					s: bytesToHex(dleq.s),
+					e: bytesToHex(dleq.e),
+					r: numberToHexPadded64(dleq.r ?? BigInt(0))
+				} as SerializedDLEQ
+			})
+		} as Proof;
```

Comment:
Instead of casting to `Proof` can we say `const serializedProof: Proof = {...`? With the way it is now, I can remove `serializeProof` and not get any type errors